packages/xo-server/src/xapi/mixins/vm.mjs

```diff
@@ -140,53 +140,100 @@ const methods = {
+        vdis = []
+      }
 
-          // if another SR is set, move it there
-          if (srId) {
-            vdi = await this.moveVdi(vdi.$id, srId)
-          }
+      forEach(existingVdis, ({ $SR: sr, ...properties }, userdevice) => {
+        const vbd = find(vm.$VBDs, { userdevice })
+        if (!vbd) {
+          return
+        }
 
-          // if the disk is bigger
-          if (size != null && size > vdi.virtual_size) {
-            await this.resizeVdi(vdi.$id, size)
-          }
-        })
-      ))
+        vdis.push({ ...properties, userdevice, sr })
+      })
+    }
 
-    // Creates the user defined VDIs.
-    //
-    // TODO: set vm.suspend_SR
-    if (!isEmpty(vdis)) {
+    if (vdis !== undefined && vdis.length > 0) {
       const devices = await this.call('VM.get_allowed_VBD_devices', vm.$ref)
-      await Promise.all(
-        mapToArray(vdis, (vdiDescription, i) =>
+      const _vdisToCreate = []
+      const _vdisToUpdate = []
+      const _vdisToDestroy = []
+
+      vdis.forEach(vdi => {
+        if (vdi.userdevice === undefined) {
+          vdi.userdevice = devices.shift()
+          _vdisToCreate.push(vdi)
+          return
+        }
+
+        // If the userdevice match no vbd, create the VDI
+        const vbd = find(vm.$VBDs, { userdevice: vdi.userdevice })
+        if (vbd === undefined) {
+          if (!vdi.destroy) {
+            const userdeviceIndex = devices.indexOf(vdi.userdevice)
+            if (userdeviceIndex === -1) {
+              throw new Error(
+                `The VDI with userdevice: ${vdi.userdevice} cannot be created. Only pass userdevice if you are sure about what you are doing`
+              )
+            }
+            delete devices[userdeviceIndex]
+            _vdisToCreate.push(vdi)
+          }
+
+          return
+        }
+
+        vdi.$ref = this.getObject(vbd.VDI).$ref
+        if (vdi.destroy) {
+          _vdisToDestroy.push(vdi)
+          return
+        }
+
+        _vdisToUpdate.push(vdi)
+      })
+
+      // TODO: set vm.suspend_SR
+      await Promise.all([
+        ..._vdisToDestroy.map(vdi => this.VDI_destroy(vdi.$ref)),
+        // Creates the user defined VDIs.
+        ..._vdisToCreate.map((vdi, i) =>
           this.VDI_create({
-            name_description: vdiDescription.name_description,
-            name_label: vdiDescription.name_label,
-            virtual_size: vdiDescription.size,
-            SR: this.getObject(vdiDescription.sr || vdiDescription.SR, 'SR').$ref,
+            name_description: vdi.name_description,
+            name_label: vdi.name_label,
+            virtual_size: vdi.size,
+            SR: this.getObject(vdi.sr, 'SR').$ref,
           }).then(vdiRef =>
             this.VBD_create({
               // Either the CD or the 1st disk is bootable (only useful for PV VMs)
               bootable: !(hasBootableDisk || i),
 
-              userdevice: devices[i],
+              userdevice: vdi.userdevice,
```

Comment:
I think the userdevice control should be kept in VBD_create, since it also handle the CDROM situation (userdevice = 3 )